File: src/sdk/ts/src/source.ts
/**
 * Catalog-source configuration seam (ADR-0003). A source is a remote catalog
 * a loader pulls skills from over the frozen protocol/v1 contract; resolution
 * here only decides *whether* a source is configured and with what
 * credentials — no loader is selected, no network is touched. Pure and
 * env-injectable, modeled on `resolveOtlpConfig` in `@ratel-ai/telemetry`.
 */

// The package compiles without node type definitions; this module-local
// declaration types the single `process.env` read (node >= 20 per `engines`).
declare const process: { env: Record<string, string | undefined> };

/** Env var naming the catalog source's base URL. Unset = the embedded floor. */
export const SOURCE_URL_ENV = "RATEL_URL";

/** Env var holding the source API key, sent as `Authorization: Bearer <key>`. */
export const SOURCE_API_KEY_ENV = "RATEL_API_KEY";

export interface CatalogSourceConfig {
  /** Base URL of the catalog source (the loader appends `/v1/catalog`). */
  url: string;
  /** API key for `Authorization: Bearer <key>` on every `/v1` request. */
  apiKey?: string;
  /** Opaque subject selector, passed through as `?scope=`. Fixed per loader. */
  scope?: string;
}

/**
 * Resolve explicit options + env into a source config, or `undefined` when no
 * URL is available anywhere — the permanent offline floor: in-process
 * registration only, byte-identical to a build with no source code paths.
 * Explicit options beat env; `scope` comes from options only.
 */
export function resolveSourceConfig(
  options?: { url?: string; apiKey?: string; scope?: string },
  env: Record<string, string | undefined> = process.env,
): CatalogSourceConfig | undefined {
  const url = options?.url ?? env[SOURCE_URL_ENV];
  if (!url) return undefined;
  const config: CatalogSourceConfig = { url };
  const apiKey = options?.apiKey ?? env[SOURCE_API_KEY_ENV];
  if (apiKey !== undefined) config.apiKey = apiKey;
  if (options?.scope !== undefined) config.scope = options.scope;
  return config;
}

File: src/sdk/python/ratel_ai/source.py
"""Catalog-source config seam — the Python mirror of `src/sdk/ts/src/source.ts`.

Pure resolution of where a catalog-source loader pulls from (ADR-0003): explicit
options beat the `RATEL_URL` / `RATEL_API_KEY` environment; no url anywhere means
no source (the permanent embedded floor). Loaders (e.g. `ratel-ai-cloud`) consume
the resolved `SourceConfig`; this module never touches the network.
"""

from __future__ import annotations

import os
from collections.abc import Mapping
from dataclasses import dataclass

__all__ = ["SOURCE_API_KEY_ENV", "SOURCE_URL_ENV", "SourceConfig", "SourceOptions"]

SOURCE_URL_ENV = "RATEL_URL"
SOURCE_API_KEY_ENV = "RATEL_API_KEY"


@dataclass(frozen=True)
class SourceOptions:
    """Explicit loader options; any field set here wins over the environment.

    `scope` is an opaque subject selector passed through as `?scope=` (ADR-0010);
    it is options-only — never read from the environment.
    """

    url: str | None = None
    api_key: str | None = None
    scope: str | None = None


@dataclass(frozen=True)
class SourceConfig:
    """A resolved catalog source: base URL, Bearer key, and fixed scope."""

    url: str
    api_key: str | None
    scope: str | None


def resolve_source_config(
    options: SourceOptions | None = None,
    env: Mapping[str, str] | None = None,
) -> SourceConfig | None:
    """Resolve options + environment into a `SourceConfig`, or `None` when no
    url is configured anywhere. Pure and env-injectable (the
    `resolve_otlp_config` pattern) so precedence is testable without patching
    `os.environ`.
    """
    opts = options or SourceOptions()
    environ = os.environ if env is None else env
    url = opts.url or environ.get(SOURCE_URL_ENV)
    if not url:
        return None
    return SourceConfig(
        url=url,
        api_key=opts.api_key or environ.get(SOURCE_API_KEY_ENV),
        scope=opts.scope,
    )
